### PR TITLE
Bed allocation quickstart

### PR DESCRIPTION
**Pull request by ge0ffrey**
- [x] domain
- [x] unit tests
- [x] integration test?
- [x] test data
- [x] UI

Optional domain improvements/suggestions:

- [x] merge the classes Stay and BedDesignation and call it Stay
- [x] Replace `Specialism` with a `String`, and `List<Specialism>` with a `List<String>` etc.
- [x] Do all classes (like bed) actually need an id?
- [x] Code style: replace "fooList" by "foos" to align with other quickstarts?

### CODE REVIEW

File: use-cases/bed-allocation/src/main/java/org/acme/bedallocation/domain/BedDesignation.java [hunk 1]
```diff
@@ -55,7 +55,7 @@ public Integer getPatientPreferredMaximumRoomCapacity() {
     }
 
     @JsonIgnore
-    public Specialism getSpecialism() {
+    public String getSpecialism() {
```
Comment (triceo) [inline use-cases/bed-allocation/src/main/java/org/acme/bedallocation/domain/BedDesignation.java]: I'm not sure how I feel about this. Two aspects:

- `Specialism` is a word I've never seen used anywhere else other than this quickstart. What is it supposed to mean, and can we find a better one?
- To me, `String` is not a data type. It carries no meaning, you cannot effectively use it in constraints. I consider this to be an anti-pattern, and I'd rather not teach our users to do this. I prefer the original approach, where this had a specific type for itself.

This has been a point where Geoffrey and I've disagreed over the years. If the `specialism` is never ever used inside the constraints, or rather used through `BedSpecialism` et al., then I can live with this. But I still don't like the name.

File: use-cases/bed-allocation/src/main/java/org/acme/bedallocation/domain/Department.java [hunk 2]
```diff
@@ -13,19 +15,20 @@ public class Department {
...
     private List<Room> rooms;
 
     public Department() {
+        this.specialismsToPriority = new HashMap<>();
     }
 
     public Department(String id, String name) {
         this.id = id;
         this.name = name;
+        this.specialismsToPriority = new HashMap<>();
```
Comment (triceo) [inline use-cases/bed-allocation/src/main/java/org/acme/bedallocation/domain/Department.java]: Careful about `HashMap`s - if you ever iterate over them, they will introduce non-determinism in your solver. However, if you ever only `get(...)` on them, it's fine.

File: use-cases/bed-allocation/src/main/java/org/acme/bedallocation/domain/DepartmentSpecialism.java [hunk 3]
```diff
@@ -8,14 +8,14 @@ public class DepartmentSpecialism {
...
-    private Specialism specialism;
+    private String specialism;
 
     private int priority; // AKA choice
 
     public DepartmentSpecialism() {
     }
 
-    public DepartmentSpecialism(String id, Department department, Specialism specialism, int priority) {
+    public DepartmentSpecialism(String id, Department department, String specialism, int priority) {
```
Comment (triceo) [inline use-cases/bed-allocation/src/main/java/org/acme/bedallocation/domain/DepartmentSpecialism.java]: I don't understand. So we do have `DepartmentSpecialism`, but for some reason, `specialism` is still a `String` and we use it as `String` everywhere?

File: use-cases/bed-allocation/src/main/java/org/acme/bedallocation/domain/Patient.java [hunk 4]
```diff
@@ -19,8 +19,8 @@ public class Patient {
     private int age;
     private Integer preferredMaximumRoomCapacity;
 
-    private List<Equipment> requiredEquipments;
-    private List<Equipment> preferredEquipments;
+    private List<String> requiredEquipments;
```
Comment (triceo) [inline use-cases/bed-allocation/src/main/java/org/acme/bedallocation/domain/Patient.java]: Same concerns on `Equipment` as on `Specialism`.

File: use-cases/bed-allocation/overconstrained01.txt [hunk 5]
```diff
@@ -0,0 +1,898 @@
+ARTICLE BENCHMARK DATA SET
```
Comment (triceo) [inline use-cases/bed-allocation/overconstrained01.txt]: Why is this file here, and should it be here?
No other quickstart carries its own data files.

File: use-cases/bed-allocation/src/main/java/org/acme/bedallocation/domain/Schedule.java [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+
+    // ************************************************************************
+    // Getters and setters
+    // ************************************************************************
+
+    @JsonIgnore
+    @ProblemFactCollectionProperty
+    public List<String> getSpecialisms() {
+        return this.departments.stream().flatMap(d -> d.getSpecialismsToPriority().keySet().stream()).distinct().toList();
+    }
```
Comment (triceo) [inline use-cases/bed-allocation/src/main/java/org/acme/bedallocation/domain/Schedule.java]: Is this ever used anywhere? I don't see it used in the constraints and if it's not in the constraints, it doesn't need to exist.

File: use-cases/bed-allocation/src/main/java/org/acme/bedallocation/domain/Specialism.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
+package org.acme.bedallocation.domain;
+
+import com.fasterxml.jackson.annotation.JsonIdentityInfo;
+import com.fasterxml.jackson.annotation.ObjectIdGenerators;
+
+import ai.timefold.solver.core.api.domain.lookup.PlanningId;
+
+@JsonIdentityInfo(scope = Specialism.class, generator = ObjectIdGenerators.PropertyGenerator.class, property = "id")
+public class Specialism  {
```
Comment (triceo) [inline use-cases/bed-allocation/src/main/java/org/acme/bedallocation/domain/Specialism.java]: If `Specialism` is now a `String`, why does this exist?

**Review by triceo — COMMENTED** [review]
Providing initial comments/questions.
Also note that you are introducing some new `pom.xml` files which shouldn't be there.

File: use-cases/bed-allocation/src/main/java/org/acme/bedallocation/domain/Schedule.java [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+import ai.timefold.solver.core.api.domain.solution.PlanningSolution;
+import ai.timefold.solver.core.api.domain.solution.ProblemFactCollectionProperty;
+import ai.timefold.solver.core.api.domain.valuerange.ValueRangeProvider;
+import ai.timefold.solver.core.api.score.buildin.hardmediumsoft.HardMediumSoftScore;
+import ai.timefold.solver.core.api.solver.SolverStatus;
+
+import com.fasterxml.jackson.annotation.JsonIgnore;
+
+@PlanningSolution
+public class Schedule {
```
Comment (triceo) [inline use-cases/bed-allocation/src/main/java/org/acme/bedallocation/domain/Schedule.java]: I'd prefer if we called this `BedPlan` or something like that.
`Schedule` says nothing, and it makes looking up this class among other quickstarts needlessly difficult.

File: use-cases/bed-allocation/src/main/java/org/acme/bedallocation/domain/BedDesignation.java (hunk 1 above)
Comment (zepfred) [reply use-cases/bed-allocation/src/main/java/org/acme/bedallocation/domain/BedDesignation.java]: I don't have a strong opinion on that. 

> I'm not sure how I feel about this. Two aspects:
> 
> * `Specialism` is a word I've never seen used anywhere else other than this quickstart. What is it supposed to mean, and can we find a better one?

During some quick research, I discovered more than one work that utilizes the term `Specialty`. What do you think?

> * To me, `String` is not a data type. It carries no meaning, you cannot effectively use it in constraints. I consider this to be an anti-pattern, and I'd rather not teach our users to do this. I prefer the original approach, where this had a specific type for itself.
> 

I see your point. I think the objective is to simplify the design of the domain models. Regardless, other quickstart modules use the same approach, e.g., `employee scheduling`.

File: use-cases/bed-allocation/src/main/java/org/acme/bedallocation/domain/Department.java (hunk 2 above)
Comment (zepfred) [reply use-cases/bed-allocation/src/main/java/org/acme/bedallocation/domain/Department.java]: We are not iterating over it. This was a solution to simplify the JSON payload.

File: use-cases/bed-allocation/src/main/java/org/acme/bedallocation/domain/DepartmentSpecialism.java (hunk 3 above)
Comment (zepfred) [reply use-cases/bed-allocation/src/main/java/org/acme/bedallocation/domain/DepartmentSpecialism.java]: That's correct. All related classes use the type `String` to define `specialism`.

File: use-cases/bed-allocation/overconstrained01.txt (hunk 5 above)
Comment (zepfred) [reply use-cases/bed-allocation/overconstrained01.txt]: No, it should not. Let me remove it.

File: use-cases/bed-allocation/src/main/java/org/acme/bedallocation/rest/DemoDataGenerator.java [hunk 9]
```diff
@@ -52,7 +52,7 @@ public class DemoDataGenerator {
     private final Random random = new Random(0);
 
     /**
-     * The dataset was generated based on the probability distributions found in the test dataset file overconstrained01.txt.
+     * The dataset was generated based on the probability distributions found in the test dataset file.
      * However, the number of patients was reduced to 40 to simplify the quickstart process.
      */
```
Comment (triceo) [inline use-cases/bed-allocation/src/main/java/org/acme/bedallocation/rest/DemoDataGenerator.java]: Maybe this entire comment doesn't apply anymore?
It references a file nobody will be able to find.

File: use-cases/bed-allocation/src/main/java/org/acme/bedallocation/domain/Schedule.java [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+    @JsonIgnore
+    @ProblemFactCollectionProperty
+    public List<DepartmentSpecialism> getDepartmentSpecialisms() {
+        return departments.stream()
+                .flatMap(d -> d.getSpecialismsToPriority().entrySet().stream()
+                        .map(e -> new DepartmentSpecialism("%s-%s".formatted(d.getId(), e.getKey()), d, e.getKey(),
+                                e.getValue()))
+                        .toList()
+                        .stream())
+                .toList();
```
Comment (triceo) [inline use-cases/bed-allocation/src/main/java/org/acme/bedallocation/domain/Schedule.java]: I'd be very careful about this pattern.

First of all - this will do some very heavy lifting every time it is called.
But more importantly, this means that every time it is called, the facts will be different instances of objects, and that will eventually lead to inefficiencies, worst case bugs.

Personally, I'd fill these fields when the departments are being set, and not touch them after.

File: use-cases/bed-allocation/src/main/java/org/acme/bedallocation/domain/Schedule.java (hunk 8 above)
Comment (ge0ffrey) [reply use-cases/bed-allocation/src/main/java/org/acme/bedallocation/domain/Schedule.java]: in a hospital, they typically call it the schedule, not the bedplan. But I don't mind if we change it. Although bedplan is a little to close to bedpan for me :)

File: use-cases/bed-allocation/src/main/java/org/acme/bedallocation/domain/BedDesignation.java (hunk 1 above)
Comment (ge0ffrey) [reply use-cases/bed-allocation/src/main/java/org/acme/bedallocation/domain/BedDesignation.java]: Just like the school timetabling quickstarts, that uses a String for Teacher etc, it makes sense for simplicity's sake to use a String for Speciality (or Specialism) and Equipment here. Keep it simple.

Comment (ge0ffrey) [reply use-cases/bed-allocation/src/main/java/org/acme/bedallocation/domain/BedDesignation.java]: Once a data type has a property beyond its name, then it should no longer be String.

**Review by triceo — CHANGES_REQUESTED** [review]
Missing the usual READMEs.
Also, the PR description has several checkboxes not checked - please clean those up so that I can be sure you did all you intended on doing.

**Review by triceo — APPROVED** [review]
